sdks/apigw-manager/src/apigw_manager/drf/authentication.py

```diff
@@ -0,0 +1,98 @@
+from typing import ClassVar, Type
+
+from apigw_manager.apigw.providers import CachePublicKeyProvider, PublicKeyProvider
+from apigw_manager.apigw.utils import get_configuration
+from django.conf import settings
+from django.contrib import auth
+from django.utils.module_loading import import_string
+from rest_framework.authentication import BaseAuthentication
+
+logger = logging.getLogger(__name__)
+
+
+class ApiGatewayJWTAuthentication(BaseAuthentication):
+    JWT_KEY_NAME = "HTTP_X_BKAPI_JWT"
+    ALGORITHM = "RS512"
+    PUBLIC_KEY_PROVIDER_CLS: ClassVar[Type[PublicKeyProvider]] = CachePublicKeyProvider
+
+    App = namedtuple("App", ["bk_app_code", "verified"])
+
+    def __init__(self):
+        configuration = get_configuration()
+        jwt_provider_cls = import_string(
+            configuration.jwt_provider_cls or "apigw_manager.apigw.providers.DefaultJWTProvider"
+        )
+        self.provider = jwt_provider_cls(
+            jwt_key_name=self.JWT_KEY_NAME,
+            default_gateway_name=configuration.gateway_name,
+            algorithm=getattr(settings, "APIGW_JWT_ALGORITHM", self.ALGORITHM),
+            allow_invalid_jwt_token=getattr(settings, "APIGW_ALLOW_INVALID_JWT_TOKEN", False),
+            public_key_provider=self.PUBLIC_KEY_PROVIDER_CLS(default_gateway_name=configuration.gateway_name),
+        )
+
+    def authenticate(self, request):
+        jwt = self.provider.provide(request)
+        if not jwt:
+            logger.error("[ApiGatewayJWTAuthentication] can not found jwt in request header")
+            return None
+
+        request.jwt = jwt
+        request._dont_enforce_csrf_checks = True
+
+        jwt_app = (jwt.payload.get("app") or {}).copy()
+        jwt_app.setdefault("bk_app_code", jwt_app.pop("app_code", None))
+
+        request.app = self.make_app(**jwt_app)
+
+        # skip when authenticated
+        # print("request attrs:", dir(request.user))
+        # if hasattr(request, "user") and request.user.is_authenticated:
+        #     return (request.user, None)
+
+        jwt_user = (jwt.payload.get("user") or {}).copy()
+        jwt_user.setdefault("bk_username", jwt_user.pop("username", None))
+
+        user = self.get_user(request, gateway_name=jwt.gateway_name, **jwt_user)
+        logger.debug("[ApiGatewayJWTAuthentication] authenticate user: %s", user)
+        return (user, None)
+
+    def authenticate_header(self, request):
+        return self.JWT_KEY_NAME
+
+    def make_app(self, bk_app_code=None, verified=False, **jwt_app):
+        return self.App(
+            bk_app_code=bk_app_code,
+            verified=verified,
+        )
+
+    def get_user(
+        self,
+        request,
+        gateway_name=None,
+        bk_username=None,
+        verified=False,
+        **credentials,
+    ):
+        # 传递 gateway_name 参数的用途：
+        # 1. 来明确标识这个请求来自于网关
+        # 2. 用户已经过认证，后端无需再认证
+        # 3. 避免非预期调用激活对应后端使得用户认证被绕过
+        return auth.authenticate(
```

Comment:
这里调 `auth.authenticate ` 会继续触发到 Django 配置的 authentication_classes 吧？但是我没有看到 Django 又继续接管这个 authentication 的类？

Reply:
会触发到 django.contrib.auth.authenticate， 然后调用对应的auth backend
```
# if you already have AUTHENTICATION_BACKENDS, delete this line
# otherwise, just keep this line
AUTHENTICATION_BACKENDS = ["django.contrib.auth.backends.ModelBackend"]

# 添加 AUTHENTICATION_BACKENDS
AUTHENTICATION_BACKENDS += [
    "apigw_manager.apigw.authentication.UserModelBackend",
]
```